pycalphad/model.py

```diff
@@ -952,84 +962,183 @@ def mole_fraction(species_name, phase_name, constituent_array,
+        disordered (with site fractions replaced by quasi mole fractions in the
+        ordered sublattices) and the interstitial sublattices will not have any
+        site fractions substituted.
+
+        References
+        ----------
+
+        .. [1] Connetable et al., Calphad 2008, 32 (2), 361–370. doi: 10.1016/j.calphad.2008.01.002
+        .. [2] Lukas, Fries, and Sundman, Computational Thermodynamics: the Calphad Method, Cambridge University Press (2007).
+
         """
         phase = dbe.phases[self.phase_name]
         ordered_phase_name = phase.model_hints.get('ordered_phase', None)
         disordered_phase_name = phase.model_hints.get('disordered_phase', None)
         if phase.name != ordered_phase_name:
             return S.Zero
-        disordered_model = self.__class__(dbe, sorted(self.components),
-                                          disordered_phase_name)
-        constituents = [sorted(set(c).intersection(self.components)) \
-                for c in dbe.phases[ordered_phase_name].constituents]
+        ordered_phase = dbe.phases[ordered_phase_name]
+        constituents = [sorted(set(c).intersection(self.components)) for c in ordered_phase.constituents]
+        disordered_phase = dbe.phases[disordered_phase_name]
+        disordered_model = self.__class__(dbe, sorted(self.components), disordered_phase_name)
+
+        # Get substitutional sublattice indices (for the ordered phase) and
+        # validate that the number of interstitial sublattices is consistent
+        # with the disordered phase.
+        # Assumes first sublattice of the disordered phase is the sublattice
+        # that can be come ordered:
+        disordered_subl_constituents = disordered_phase.constituents[0]
+        ordered_constituents = ordered_phase.constituents
+        substitutional_sublattice_idxs = []
+        for idx, subl_constituents in enumerate(ordered_constituents):
+            # Assumes that the ordered phase sublattice describes the ordering
+            # if it has exactly the same constituents. Could be a source of
+            # false positives if any interstitial sublattices have the same
+            # constituents as the disordered sublattice, but there's not an
+            # explicit way to specify which sublattices are ordering. We try to
+            # compensate for this assumption by validating (next).
+            if len(disordered_subl_constituents.symmetric_difference(subl_constituents)) == 0:
+                substitutional_sublattice_idxs.append(idx)
+        # validate
+        num_substitutional_sublattice_idxs = len(substitutional_sublattice_idxs)
+        num_ordered_interstitial_subls = len(ordered_phase.sublattices) - num_substitutional_sublattice_idxs
+        num_disordered_interstitial_subls = len(disordered_phase.sublattices) - 1
+        if num_ordered_interstitial_subls != num_disordered_interstitial_subls:
+            raise ValueError(
```

Comment:
Is there an easy test which could be written for this?

Reply:
Added in c5d0852